INT-3736: Process Barrier Component

garyrussell opened the pull request:
JIRA: https://jira.spring.io/browse/INT-3736

Needs docs and javadoc

garyrussell committed:
INT-3736: Process Barrier Component

JIRA: https://jira.spring.io/browse/INT-3736

artembilan commented:
> * What went wrong:
Execution failed for task ':spring-integration-core:checkTestConfigs'.

> Hardcoded XSD version in the config files:
  spring-integration-core/src/test/java/org/springframework/integration/config/xml/BarrierParserTests-context.xml
  Please, use versionless schemaLocations for Spring XSDs to avoid issues with builds on different versions of dependencies.

Ts-ts-ts :smile: ...

garyrussell committed:
INT-3736: Polishing

garyrussell commented:
Pushed, with polish

garyrussell committed:
INT-3736: Use o-c-a For Release

Avoids unusual parser logic, allows release channel to be of any type.

garyrussell commented:
Pushed - using o-c-a for release - much simpler and flexible - WDYT?

artembilan commented on spring-integration-core/src/main/resources/org/springframework/integration/config/xml/spring-integration-4.2.xsd:
```diff
@@ -1659,6 +1659,80 @@
+					<xsd:complexType>
+						<xsd:complexContent>
+							<xsd:extension base="handlerAdviceChainType">
+								<xsd:attribute name="discard-within-advice" type="xsd:string" default="true">
```
What does this feature do here ?

garyrussell replied:
Doh - copy/paste

garyrussell committed:
INT-3736: Schema Polishing

garyrussell commented:
Pushed

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/handler/SuspendingMessageHandler.java:
```diff
@@ -0,0 +1,132 @@
+
+/**
+ * A pair of message handlers; the first suspends the thread until the second receives
+ * a corresponding message.
```
This is bad JavaDocs now, but just keep in mind: no need to be in a rush.
I let you know when I finish. review.

artembilan commented on spring-integration-core/src/test/java/org/springframework/integration/config/xml/BarrierParserTests-context.xml:
```diff
@@ -0,0 +1,31 @@
+
+	<int:channel id="release" />
+
+	<int:outbound-channel-adapter channel="release" ref="barrier1.handler" method="release" />
```
I feel I'm fine with the solution.
Even if some one doesn't like to use `<outbound-channel-adapter>` on the matter, he can `inject` that `MessageHandler` to invoke the `release()` directly.
Although we must think about extraction some interface on the matter to cover the `Proxy` case.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/xml/AbstractCorrelatingMessageHandlerParser.java:
```diff
@@ -63,10 +60,10 @@
 	protected void doParse(BeanDefinitionBuilder builder, Element element, BeanMetadataElement processor,
 			ParserContext parserContext) {
-		this.injectPropertyWithAdapter(CORRELATION_STRATEGY_REF_ATTRIBUTE, CORRELATION_STRATEGY_METHOD_ATTRIBUTE,
+		IntegrationNamespaceUtils.injectPropertyWithAdapter(CORRELATION_STRATEGY_REF_ATTRIBUTE, CORRELATION_STRATEGY_METHOD_ATTRIBUTE,
```
These are changes to the class, so update `Copyright`!
Minor, of course, and can be addressed on merge.

artembilan replied:
Plus `@author`

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/xml/BarrierParser.java:
```diff
@@ -0,0 +1,43 @@
+import org.springframework.integration.handler.SuspendingMessageHandler;
+
+/**
+ * @author Gary Russell
```
Need some short JavaDoc.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/handler/SuspendingMessageHandler.java:
```diff
@@ -0,0 +1,132 @@
+ * @since 4.2
+ *
+ */
+public class SuspendingMessageHandler extends AbstractReplyProducingMessageHandler {
```
Well, even if `Suspending` word is more closer to the implementation of the class, but in face of Java Config, I'd prefer now to go ahead with symmetrical names for XML component, as well as for their classes on the background.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/handler/SuspendingMessageHandler.java:
```diff
@@ -0,0 +1,132 @@
+	private final ConcurrentHashMap<Object, BlockingQueue<Message<?>>> suspensions =
+			new ConcurrentHashMap<Object, BlockingQueue<Message<?>>>();
+
+	private final ConcurrentHashMap<Object, Thread> inProcess = new ConcurrentHashMap<Object, Thread>();
```
Both variables must in the `interface` types.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/handler/SuspendingMessageHandler.java:
```diff
@@ -0,0 +1,132 @@
+
+	@Override
+	protected Object handleRequestMessage(Message<?> requestMessage) {
+		Object key = SuspendingMessageHandler.this.correlationStrategy.getCorrelationKey(requestMessage);
```
Redundant class reference

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/handler/SuspendingMessageHandler.java:
```diff
@@ -0,0 +1,132 @@
+				}
+			}
+			else {
+				synchronized(blockingQueue) {
```
IDEA says me that this `synchronized` is redundant, because of **local** variable.

garyrussell replied:
Yes, this was there because I was previously doing more work in the block, but now it's down to a simple `offer` in both places.

garyrussell replied:
That's actually incorrect (IDEA) because the local variable is populated from a field.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/handler/SuspendingMessageHandler.java:
```diff
@@ -0,0 +1,132 @@
+				synchronized(blockingQueue) {
+					if (!blockingQueue.offer(requestMessage)) {
+						// second chance
+						return buildResult(requestMessage, blockingQueue.poll());
```
I like this tolerance, but `this.suspensions.remove(key);` is missed in this case anyway :smile: 
Plus you don't check the result type, too...
In other words: this `second chance` must be the same by the algorithm with the general code block.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/handler/SuspendingMessageHandler.java:
```diff
@@ -0,0 +1,132 @@
+						return buildResult(requestMessage, blockingQueue.poll());
+					}
+				}
+				return null;
```
I think just return `null` isn't enough for this component.
How about `MessageTimeoutException`?
From other side we can regulate the final behavior with some options. E.g. `requiresReply`.
WDYT?

garyrussell replied:
Right - I assumed `requiresReply` would control that.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/handler/SuspendingMessageHandler.java:
```diff
@@ -0,0 +1,132 @@
+		Object[] result = new Object[2];
+		result[0] = requestMessage.getPayload();
+		result[1] = release;
+		return result;
```
I think we must provide the hook to inject `MessageListProcessor` to allow to decide what to do both those messages. 
By default we can do the same what `AggregatingMessageHandler` does - `DefaultAggregatingMessageGroupProcessor`.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/handler/SuspendingMessageHandler.java:
```diff
@@ -0,0 +1,132 @@
+			if (!blockingQueue.offer(message)) {
+				this.logger.error("Release message arrived too late: " + message
+						+ " pending message: " + blockingQueue.poll());
+				this.suspensions.remove(key);
```
OK. I see now your point regarding that `second chance`.
Enough smart, though...
But seems for me we can end up with `OOM`, if we allow to retain unreleased/cleared `suspensions`.
I mean that case, when the `releaseMessage` won't be sent at all.

Your thoughts ?

garyrussell replied:
I think I can resolve that, now that we have the `inProcess` field.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/handler/SuspendingMessageHandler.java:
```diff
@@ -0,0 +1,132 @@
+			}
+			else {
+				synchronized(blockingQueue) {
+					if (!blockingQueue.offer(requestMessage)) {
```
So, in case of timeout in the main block we add the `requestMessage` to the `suspensions` taking the queue for the `correlationKey` and remove `inProcess` for it.
In addition we return `null` to the main flow to discard the message by the `requiresReply` reason or some other logic. Doesn't matter already for us.
My concern is here that the user eventually decides to send the message with the same `correlationKey`.
The `inProcess` is free anyway, the `createOrObtainQueue` returns good existing `queue` and we go ahead to the `poll()`.
And here is an underwater rock: the `poll()` returns for us the previous outdated `requestMessage`.
Voila!

I don't see the solution yet, but at least you should pay attention...

artembilan commented on spring-integration-core/src/main/resources/org/springframework/integration/config/xml/spring-integration-4.2.xsd:
```diff
@@ -1659,6 +1659,63 @@
+	<xsd:element name="barrier">
+		<xsd:annotation>
+			<xsd:documentation>
+				Defines an endpoint that suspends a thread until a corresponding release message is received.
```
I think that would be good to mention here the target class for the component.
Seems for me it should help  a bit users who migrate from the XML configuration to DSL.

According to that `Pitch Black` of questions on SO, it looks like people just don't know which classes are related to their XML config.

I don't say that we must do that for everything just in place, but at least we can start do that for new components.

artembilan commented on spring-integration-core/src/main/resources/org/springframework/integration/config/xml/spring-integration-4.2.xsd:
```diff
@@ -1659,6 +1659,63 @@
+			<xsd:attribute name="timeout" type="xsd:string" use="required">
+				<xsd:annotation>
+					<xsd:documentation>
+						The time in milliseconds to suspend the thread. See `requires-reply`.
```
Just as reminder - the `requires-reply` must be exposed.

artembilan commented on spring-integration-core/src/test/java/org/springframework/integration/config/xml/BarrierParserTests.java:
```diff
@@ -0,0 +1,78 @@
+		assertEquals(10000L, TestUtils.getPropertyValue(handler, "timeout"));
+
+		assertThat(TestUtils.getPropertyValue(this.barrier2, "handler.correlationStrategy"),
+				instanceOf(HeaderAttributeCorrelationStrategy.class));
```
Just as nit-picking: not enough tests for parser logic.
Right, we can live with this as well, but seems for me we have catched some issues with lack of coverage some time ago...

garyrussell replied:
? I don't think it's necessary to test the properties from the abstract parser; I think everything new is covered here.

artembilan replied:
Well, I don't see `correlation-strategy-expression` test for the `barrier1`

garyrussell replied:
The runtime test tests that (`'foo'`) :smile:

artembilan replied:
:wink: OK. Never mind.
I don't feel good today: looks like I've catched a cold...

artembilan commented on spring-integration-core/src/test/java/org/springframework/integration/handler/SuspendingMessageHandlerTests.java:
```diff
@@ -0,0 +1,197 @@
+		assertSame(exc, exception.get().getCause());
+		assertEquals(0, suspensions.size());
+	}
+
```
... and need the test to cover my use-case for the second `requestMessage` after `timeout`.

artembilan commented:
That's all for this cycle.

garyrussell committed:
INT-3736: Polishing

- Add TriggerMessageHandler interface
- Use SyncQueue - suspend the trigger thread if the main thread hasn't arrived yet
 - allows clean up of state regardless of whether just a trigger or suspend is processed
- reply-required

garyrussell commented:
Pushed

- Add TriggerMessageHandler interface
- Use `SyncQueue` - suspend the trigger thread if the main thread hasn't arrived yet
 - allows clean up of state regardless of whether just a trigger or suspend is processed
- reply-required

garyrussell commented:
Re-pushed polish to fix javadocs

artembilan commented on spring-integration-core/src/main/resources/org/springframework/integration/config/xml/spring-integration-4.2.xsd:
```diff
@@ -1659,6 +1659,91 @@
+					</xsd:documentation>
+				</xsd:annotation>
+			</xsd:attribute>
+			<xsd:attribute name="output-processor" type="xsd:string">
```
Don't you like an expression support as an alternative here?

garyrussell replied:
I thought about it, but it's not likely it is something people will want to change; and they can always wire in an `ExpressionEvaluating`...

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/handler/BarrierMessageHandler.java:
```diff
@@ -0,0 +1,192 @@
+	}
+
+	private Object processRelease(Message<?> requestMessage, Object key, Message<?> release) {
+		this.suspensions.remove(key);
```
Or this one or that in the `finally` above is redundant.
Although for me this one is better, because we are in front of sending reply, so the `key` is really free already to suspend a new pair.

garyrussell replied:
Agreed, but it still needs to be in the finally block in case of timeout or interrupt.

garyrussell replied:
Oh, no - it might remove the next one - I think we can only use the one in the finally block

garyrussell replied:
No; it's ok; we're protected by inProcess - but effectively, this one does not free up the key earlier because of `inProcess` so the conclusion is remove this one, I think.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/handler/TriggerMessageHandler.java:
```diff
@@ -0,0 +1,36 @@
+ * @since 4.2
+ *
+ */
+public interface TriggerMessageHandler {
```
How about `TriggerActionAware`, or `TriggerAction`?
It isn't `MessageHandler` anyway...

garyrussell replied:
`MessageTriggerable`, `MessageTriggerAction` ??

artembilan replied:
Yes, I like `MessageTriggerAction`! :+1: 
I remember when I worked with Struts 2 and the `Action` with `execute()` was good contract to implement from there.
So, we really can make end-users to live with our `MessageTriggerAction` and just `trigger()` method from it.
WDYT?

garyrussell replied:
:+1:

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/handler/BarrierMessageHandler.java:
```diff
@@ -0,0 +1,192 @@
+	 * @param release the release message.
+	 * @return the result.
+	 */
+	protected Object buildResult(Object key, Message<?> requestMessage, Message<?> release) {
```
`releaseMessage` sounds better.
But yes: it's minor and just my whimsy

garyrussell replied:
ok; if you want to change during merge.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/handler/BarrierMessageHandler.java:
```diff
@@ -0,0 +1,192 @@
+		}
+		SynchronousQueue<Message<?>> syncQueue = createOrObtainQueue(key);
+		try {
+			if (syncQueue != null && !syncQueue.offer(message, timeout, TimeUnit.MILLISECONDS)) {
```
To be honest I don't like this `blocking` solution.
The `trigger` Thread can't depend on the main Thread.
And in my opinion the `triggerAction` for non-existing `key` must be rejected with the `MessageRejectedException`.

In addition: how about to consider to use `LockRegistry` for the `key` to suspend other `Thread` instead of fail fast on the `inProcess`?

garyrussell replied:
If we don't block we'd need a reaper to clean up orphans.

Let's go with this; we can always improve later if people complain.

99/100 the release thread will arrive last.

garyrussell replied:
> MessageRejectedException

We can't do that - the trigger thread might arrive before the suspending thread.

garyrussell replied:
But, yes, we can use LockRegistry - good idea.

garyrussell replied:
Actually - I don't think this is a good idea after all.

If we allow concurrent use of the same correlation id, it is indeterminate as to which trigger will arrive first.

I think it's safer to fail fast.

artembilan replied:
If you mean `inProcess` monitor, I agree.
The `LockRegistry` would not be so flexible there.
I just don't understand yet what you are going to do here on the early trigger action...

If we don't allow concurrent suspend (`inProcess` monitor must help), we should allow concurrent `triggerAction` as well.
Have I missed anything?

artembilan commented:
Everything else is OK, unless we'll decide to rework to other algorithm, hence you will be forced to rework tests..

garyrussell commented:
Oh - I need to do docs anyway - I'll make these changes with that commit

garyrussell commented:
Will do final polish and docs.

garyrussell committed:
INT-3736: Polishing and Docs

garyrussell commented:
Rebased, pushed.

artembilan commented on src/reference/asciidoc/barrier.adoc:
```diff
@@ -0,0 +1,87 @@
+[[barrier]]
+=== Thread Barrier
+
+Sometimes, we need to suspend a message flow thread until some other asynchronous even occurs.
```
Typo: even**t**
Will be fixed on merge.
With other simple code style polishing.

artembilan commented on src/reference/asciidoc/barrier.adoc:
```diff
@@ -0,0 +1,87 @@
+We might wish to not reply to the user until the RabbitMQ broker has issued an acknowledgment that the message was
+received.
+
+_version 4.2_ introduced the `<barrier/>` component for this purpose.
```
I think would be better to have more full expression here for target readers:
> Since _version 4.2_ Spring Integration introduced...

Fixing on merge...

artembilan commented on src/reference/asciidoc/barrier.adoc:
```diff
@@ -0,0 +1,87 @@
+caller.
+When `true`, a `ReplyRequiredException` is thrown.
+
+You can call the `trigger()` method programmatically (obtain the bean reference using the name `barrier.hander`
```
Typo: hand**l**er.
And I'd prefer to distinguish `barrier` word on the next line.

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/c695129bd5efb3061e20a63f74a726628f168334
